src/hotspot/share/opto/gcm.cpp

```diff
@@ -662,23 +663,65 @@ class DefUseMemStatesQueue : public StackObj {
+// starts at the load's early block (computed in schedule_early) and ends at
+// the argument LCA block. However, there may still exist anti-dependent stores
+// in between the early block and the LCA that overwrite memory that the load
+// must witness. For such stores, we must
 //
-// Do not add edges to stores on distinct control-flow paths;
-// only add edges to stores which might interfere.
+//   1. raise the load's LCA to force the load to (eventually) be scheduled at
+//      latest in the store's block, and
+//   2. if the load may get scheduled in the store's block, additionally insert
+//      an anti-dependence edge from the load to the store to ensure LCM
+//      schedules the load before the store within the block.
 //
-// Return the (updated) LCA.  There will not be any possibly interfering
-// store between the load's "early block" and the updated LCA.
-// Any stores in the updated LCA will have new precedence edges
-// back to the load.  The caller is expected to schedule the load
-// in the LCA, in which case the precedence edges will make LCM
-// preserve anti-dependences.  The caller may also hoist the load
-// above the LCA, if it is not the early block.
-Block* PhaseCFG::insert_anti_dependences(Block* LCA, Node* load, bool verify) {
+// For a given store, we say that the store is on a _distinct_ control-flow
+// path relative to the load if there are no paths from early to LCA that go
+// through the store's block. Such stores are not anti-dependent, and there is
+// no need to update the LCA nor to add anti-dependence edges.
+//
+// Due to the presence of loops, we must also raise the LCA above
+// anti-dependent memory Phis. We defer the details (see later comments in the
+// method) and for now look at an example without loops.
+//
+//          CFG               DOMINATOR TREE
+//
+//       B1 (early,L)              B1
+//       |\________                /\\___
+//       |         \              /  \   \
+//       B2 (L,S)   \            B2  B7  B6
+//      /  \         \           /\\___
+//     B3  B4 (S)    B7 (S)     /  \   \
+//      \  /         /         B3  B4  B5
+//       B5 (LCA,L) /
+//        \    ____/
+//         \  /
+//          B6
+//
+// Here, the load's scheduling range when calling raise_above_anti_dependences
+// is between early and LCA in the dominator tree, i.e., in block B1, B2, or B5
+// (indicated with "L"). However, there are a number of stores (indicated with
+// "S") that overwrite the memory which the load must witness. First, consider
+// the store in B4. We cannot legally schedule the load in B4, so an
+// anti-dependence edge is redundant. However, we must raise the LCA above
+// B4, which means that the updated LCA is B2. Now, consider the store in B2.
+// Raising the LCA above B2 has no effect, because B2 is on the dominator tree
+// branch between early and the current LCA (in fact, B2 is the current LCA).
```

Comment:
I found this sentence a bit unclear, could you clarify what you mean by "has no effect"?

Reply:
Now simplified, does it look better? What I meant is that we do not need to raise the LCA because it is already high enough.

Reply:
> Now simplified, does it look better?

Yes, thanks!